pkg/graphite/convert/whisperconverter/whisper.go

```diff
@@ -54,53 +54,72 @@ func ReadPoints(w Archive, name string) ([]whisper.Point, error) {
-
-	// Preallocate space for all allPoints in one slice.
-	allPoints := make([]pointWithPrecision, totalPoints)
-	pIdx := 0
 	// Dump one precision level at a time and write into the output slice.
+	// Its important to remember that the archive with index 0 (first archive)
+	// has the raw data and the highest precision https://graphite.readthedocs.io/en/latest/whisper.html#archives-retention-and-precision
+	seenTs := map[uint32]struct{}{}
+	lastMinTs := uint32(0)
+	var keptPoints []whisper.Point
 	for i, a := range w.GetArchives() {
 		archivePoints, err := w.DumpArchive(i)
 		if err != nil {
 			return nil, fmt.Errorf("failed to dump archive %d from whisper metric %s", i, name)
 		}
-		for j, p := range archivePoints {
-			allPoints[pIdx+j] = pointWithPrecision{p, a.SecondsPerPoint}
+
+		var minArchiveTs, maxArchiveTs uint32
+		// calculate maxArchiveTs
+		for _, p := range archivePoints {
+			// We want to track the max timestamp of the archive because we know
+			// it virtually represents now() and we wont have newer points.
+			// Then the min timestamp of the archive would be maxTs - the archive
+			// retention.
+			if p.Timestamp > maxArchiveTs {
+				maxArchiveTs = p.Timestamp
+			}
 		}
-		pIdx += len(archivePoints)
-	}
 
-	// Points must be in time order.
-	sort.Slice(allPoints, func(i, j int) bool {
-		return allPoints[i].Timestamp < allPoints[j].Timestamp
-	})
+		// calculate minArchiveTs
+		if maxArchiveTs-a.Retention() > maxArchiveTs { // overflow, very big retention
+			minArchiveTs = 0
+		} else {
+			minArchiveTs = maxArchiveTs - a.Retention()
+		}
 
-	trimmedPoints := []whisper.Point{}
-	for i := 0; i < len(allPoints); i++ {
-		// Remove all points of time = 0.
-		if allPoints[i].Timestamp == 0 {
-			continue
+		// For subsequent archives
+		if lastMinTs > 0 {
+			// if we are already in the second or subsequent archive and we had
+			// some points in the prior archives, we want to skip
+			// samples in the previous archives
+			maxArchiveTs = lastMinTs
 		}
-		// There might be duplicate timestamps in different archives. Take the
-		// higher-precision archive value since it's unaggregated.
-		if i > 0 && allPoints[i].Timestamp == allPoints[i-1].Timestamp {
-			if allPoints[i].secondsPerPoint == allPoints[i-1].secondsPerPoint {
-				return nil, fmt.Errorf("duplicate timestamp at same precision in archive %s: %d", name, allPoints[i].Timestamp)
+		lastMinTs = minArchiveTs
+
+		for _, p := range archivePoints {
+			if p.Timestamp < minArchiveTs {
+				continue
+			}
+			// If we have already seen a point with the same timestamp it means
+			// we already have a point from an archive with higher precision that
+			// we want to keep. So we skip this point.
+			if _, ok := seenTs[p.Timestamp]; ok {
+				continue
 			}
-			if allPoints[i].secondsPerPoint < allPoints[i-1].secondsPerPoint {
-				trimmedPoints[len(trimmedPoints)-1] = allPoints[i].Point
+			// Skip points with time = 0
+			if p.Timestamp == 0 {
+				continue
 			}
```

Comment:
I'd move this check to the of `for` loop, before we do other checks on `p.Timestamp`.